### ISSUE TITLE
EVP_VerifyFinal() failed with EC cert on 0.6.11

### ISSUE THREAD

**Issue by lgarbarini** [OpenSC/pam_pkcs11#44]
Hello, I have pam_pkcs11 working well with RSA certificates, but for a variety of reasons I'd much prefer EC certs, so I upgraded my pam_pkcs11 to 0.6.11 for its greater support. However, when I sign and store a ECDSA certificate, I get the following:

```
gdm-password]: loading pkcs #11 module...
gdm-password]: PKCS #11 module = [/usr/lib/opensc-pkcs11.so]
gdm-password]: module permissions: uid = 0, gid = 0, mode = 755
gdm-password]: loading module /usr/lib/opensc-pkcs11.so
gdm-password]: getting function list
gdm-password]: initialising pkcs #11 module...
gdm-password]: module information:
gdm-password]: - version: 2.20
gdm-password]: - manufacturer: OpenSC Project
gdm-password]: - flags: 0000
gdm-password]: - library description: OpenSC smartcard framework
gdm-password]: - library version: 0.20
gdm-password]: number of slots (a): 1
gdm-password]: number of slots (b): 1
gdm-password]: slot 1:
gdm-password]: - description: Alcor Micro AU9560 00 00
gdm-password]: - manufacturer: Generic
gdm-password]: - flags: 0007
gdm-password]: - token:
gdm-password]:   - label: MyEID (Basic PIN)
gdm-password]:   - manufacturer: Aventra Ltd.
gdm-password]:   - model: PKCS#15
gdm-password]:   - serial: 7303016809988479
gdm-password]:   - flags: 040d
gdm-password]: opening a new PKCS #11 session for slot 1
gdm-password]: login as user CKU_USER
gdm-password]: Saving Certificate #1:
gdm-password]: - type: 00
gdm-password]: - id:   02
gdm-password]: Found 1 certificates in token
gdm-password]: Retrieveing mapper module list
gdm-password]: Loading static module for mapper 'ms'
gdm-password]: Inserting mapper [ms] into list
gdm-password]: verifying the certificate #1
gdm-password]: Neither CA nor CRL check requested. CertVrfy() skipped
gdm-password]: Mapper module ms match() returns 1
gdm-password]: certificate is valid and matches the user
gdm-password]: reading 128 random bytes from /dev/urandom
gdm-password]: random-value[128] = [...]
gdm-password]: private key type: 0x00000003
gdm-password]: hash[51] = [...]
gdm-password]: signature[64] = [...]
gdm-password]: verifying signature...
gdm-password]: public key type: 0x00000198
gdm-password]: public key bits: 0x00000100
gdm-password]: hashing with SHA256
gdm-password]: logout user
gdm-password]: closing the PKCS #11 session
gdm-password]: releasing keys and certificates
gdm-password]: verify_signature() failed: EVP_VerifyFinal() failed: error:25066067:DSO support routines:dlfcn_load
```

Output of: `pkcs11-tool -O`
```
Using slot 0 with a present token (0x0)
Certificate Object; type = X.509 cert
  label:       [Removed]
  subject:    [Removed]
  ID:         02
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:    [Removed]
  EC_PARAMS:   [Removed]
  label:       [Removed]
  ID:         02
  Usage:      encrypt, verify
  Access:     local
```

Output of: `pkcs11-tool -M`

```
Using slot 0 with a present token (0x0)
Supported mechanisms:
  SHA-1, digest
  SHA224, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
  MD5, digest
  RIPEMD160, digest
  GOSTR3411, digest
  ECDSA, keySize={192,256}, hw, sign, other flags=0x1800000
  ECDSA-SHA1, keySize={192,256}, hw, sign, other flags=0x1800000
  ECDH1-COFACTOR-DERIVE, keySize={192,256}, hw, derive, other flags=0x1800000
  ECDH1-DERIVE, keySize={192,256}, hw, derive, other flags=0x1800000
  ECDSA-KEY-PAIR-GEN, keySize={192,256}, hw, generate_key_pair, other flags=0x1800000
  RSA-X-509, keySize={512,2048}, hw, decrypt, sign, verify
  RSA-PKCS, keySize={512,2048}, hw, decrypt, sign, verify
  SHA1-RSA-PKCS, keySize={512,2048}, sign, verify
  RSA-PKCS-PSS, keySize={512,2048}, hw, sign, verify
  SHA1-RSA-PKCS-PSS, keySize={512,2048}, sign, verify
  RSA-PKCS-KEY-PAIR-GEN, keySize={512,2048}, generate_key_pair
  AES-ECB, keySize={128,256}, encrypt, decrypt
  AES-CBC, keySize={128,256}, encrypt, decrypt
  AES-CBC-PAD, keySize={128,256}, encrypt, decrypt
```

**Comment by lgarbarini** [comment]
This seems to be an issue with the ASN1 encoding of the ECDSA signature. I hacked together some code by following the information in [How can I convert a DER ECDSA signature to ASN.1](https://crypto.stackexchange.com/questions/1795/how-can-i-convert-a-der-ecdsa-signature-to-asn-1/1797) and can now successfully verify ECDSA signatures. 

I'm unfamiliar with these OpenSSL functions so I was having trouble understanding the original code, but one of the primary issues seems to be "Signed big-endian encoding of minimal length", from the stackexchange post:

> "Signed big-endian encoding of minimal length" means that the numerical value must be encoded as a sequence of bytes, such that the least significant byte comes last (that's what "big endian" means), the total length is the shortest possible to represent the value (that's "minimal length"), and the first bit of the first byte specifies the sign of the value (that's "signed"). For ECDSA, the r and s values are positive integers, so the first bit of the first byte must be a 0; i.e. the first byte of `(vr)` (respectively `(vs)`) must have a value between 0x00 and 0x7F.

This is my first foray into `pam_pkcs11` so I'll probably have to take some time to write a clean pull request. But for now, here is the "hacked" function:

```
  if (EVP_PKEY_base_id(pubkey) == EVP_PKEY_EC) {
    rs_len = *signature_length / 2;
    `unsigned` char* old = *signature;
    // memcpy(, old+rs_len, rs_len);

    int extend_r = 0;
    int extend_s = 0;

    // first byte of (vr) would be read as negative number, remember to pad
    if ((unsigned char)*old > 127) {
      extend_r = 1;
    }
    // first byte of (vs) would be read as negative number, remember to pad
    if (((unsigned char)*(old + rs_len)) > 127) {
      extend_s = 1;
    }

    // new signature length is 0x30 b1 0x02 b2 (vr) 0x02 b3 (vs), plus padding 
    *signature_length = (rs_len * 2) + 6 + extend_r + extend_s;

    // use calloc to handle 0 padding
    *signature = calloc(*signature_length, sizeof(char*));
    void * ptr = *signature;
    DBG1("length is: %d",(*signature_length));
    memset(ptr, 0x30, 1);
    
    // single byte length of all fields after this one
    ptr += 1;
    memset(ptr, *signature_length - 2, 1);

    // marker
    ptr += 1;
    memset(ptr, 0x02, 1);

    // length of (vr), include padding (if required)
    ptr += 1;
    memset(ptr, rs_len + extend_r, 1);

    // vr (padding handled)
    ptr += 1 + extend_r;
    memcpy(ptr, old, rs_len);
    
    // marker
    ptr += rs_len;
    memset(ptr, 0x02, 1);
    
    // length of (vs), include padding (if required)
    ptr += 1;
    memset(ptr, rs_len + extend_s, 1);
    
    // vs (padding handled)
    ptr += 1 + extend_s;
    memcpy(ptr, old+rs_len, rs_len);
```

**Comment by thyagarajan-balakrishnan** [comment]
I have the same problem and I have EC key secp521r1 in my token. This patch didn't work as is but tweaking it a bit did the trick.

    *signature_length = (rs_len * 2) + 6 + extend_r + extend_s + 1;
    ....
    memset(ptr, 0x30, 1);

    // set 0x81 after 0x30 - without this for EC521 key verification fails
    ptr += 1;
    memset(ptr, 0x81, 1);

    ptr += 1;
    memset(ptr, *signature_length - 3, 1);

**Comment by popovec** [comment]
IMHO this issue is already fixed by pull request #51 ..  (try the  https://github.com/OpenSC/pam_pkcs11/pull/51/commits/75fe90e29b35593a6a07ffc77b7f26bd611bf5a0 instead of the patch listed above)

**Comment by thyagarajan-balakrishnan** [comment]
> IMHO this issue is already fixed by pull request #51 .. (try the [75fe90e](https://github.com/OpenSC/pam_pkcs11/commit/75fe90e29b35593a6a07ffc77b7f26bd611bf5a0) instead of the patch listed above)

Tested #51, and it works for EC-521. Thanks. :+1:

**Comment by wolneykien** [comment]
#51 merged into master.
